power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -59,20 +60,63 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
 
 template <symmetry_tag sym>
 inline void calculate_source_result(IdxRange const& sources, Idx bus_number, YBus<sym> const& y_bus,
-                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output) {
-    for (Idx const source : sources) {
-        ComplexValue<sym> const u_ref{input.source[source]};
-        ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
-        output.source[source].i = dot(y_ref, u_ref - output.u[bus_number]);
-        output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
+                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
+                                    ComplexValue<sym> const& i_load_gen_bus) {
+    ComplexValue<sym> const i_source_total =
+        conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
+    std::vector<Idx> sources_acc(sources.size());
+    std::ranges::transform(sources, sources_acc.begin(), [&](Idx const source) { return source; });
+    if (sources_acc.size() == 1) {
```

Comment:
The size=0 goes via other condition.
```suggestion
    if (sources_acc.size() < 2) {
```
Or better reverse the condition for early return.

Reply:
size=0 is possible? :O

Reply:
Addressed by explicitly taking care of the size=0 case. Thanks!